### PR TITLE
sluice v2

### PR DESCRIPTION
**Pull request by galenxing**
this one

### CODE REVIEW

File: deepchem/metrics/__init__.py [hunk 1]
```diff
@@ -181,7 +181,7 @@ def compute_metric(self,
                      y_true,
                      y_pred,
                      w=None,
-                     n_classes=2,
+                     n_classes=20,
```
Comment (lilleswing) [inline deepchem/metrics/__init__.py:184]: why did you need this?

File: deepchem/models/tensorgraph/layers.py [hunk 2]
```diff
@@ -140,6 +140,129 @@ def convert_to_layers(in_layers):
...
+            raise ValueError("AlphaShare must have more than one input")
+
+        # create subspaces
+        subspaces = []
+        original_cols = int(inputs[0].get_shape()[-1].value)
+        subspace_size = int(original_cols / 2)
+        for input_tensor in inputs:
+            subspaces.append(tf.reshape(input_tensor[:, :subspace_size], [-1]))
+            subspaces.append(tf.reshape(input_tensor[:, subspace_size:], [-1]))
+        n_alphas = len(inputs) * 2
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: len(subspaces) is more understandable

File: deepchem/models/tensorgraph/layers.py [hunk 3]
```diff
@@ -140,6 +140,129 @@ def convert_to_layers(in_layers):
...
+        for x in range(0, len(inputs)):
+            lin_comb.append(tf.reshape(
+                out_tensor[row:row + 2, ], [-1, original_cols]))
+            row += 2
+        out_tensor = tf.stack(lin_comb)
+        self.alphas = alphas
+        if set_tensors:
+            self.out_tensor = out_tensor
+        return out_tensor
+
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:196]: p.s. you also need to implement none_tensors and set_tensors if you set class variables that are tensor objects.

If you don't these layers will not be save-able

File: deepchem/models/tensorgraph/layers.py [hunk 4]
```diff
@@ -140,6 +140,129 @@ def convert_to_layers(in_layers):
...
+        tower_num: int
+            corresponds to the order that layers were inputted into
+            AlphaShare layer
+        """
+        self.tower_num = tower_num
+        super(LayerSplitter, self).__init__(**kwargs)
+
+    def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+        inputs = self._get_input_tensors(in_layers)[0]
+        self.out_tensor = inputs[self.tower_num, :]
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:227]: ??? isn't inputs a list?

File: examples/SluiceTest/sluice_tester.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        y_hot[row, value] = 1
+    return y_hot
+
+
+def graph_conv_model(batch_size, tasks):
+    model = TensorGraph(model_dir=model_dir,
+                        batch_size=batch_size, use_queue=False)
+    sluice_cost = []
+
+    X1 = Feature(shape=(None, 10))
```
Comment (lilleswing) [inline examples/SluiceTest/sluice_tester.py:39]: 10?  This isn't using the data from sluice_datasets.

File: examples/SluiceTest/sluice_tester.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+        labels.append(label)
+        cost = SoftMaxCrossEntropy(in_layers=[label, classification])
+        costs.append(cost)
+
+    s_cost = SluiceLoss(in_layers=sluice_cost)
+    entropy = Concat(in_layers=costs)
+    task_weights = Weights(shape=(None, len(tasks)))
+    loss = WeightedError(in_layers=[entropy, task_weights])
+    loss = Add(in_layers=[loss, s_cost])
```
Comment (lilleswing) [inline examples/SluiceTest/sluice_tester.py:96]: for the training set total loss what is the distribution between weighted softmax CE and sluiceloss?

File: examples/SluiceTest/sluice_tester.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+    model.set_alphas([as1, as2])
+    model.set_betas([b1, b2])
```
Comment (lilleswing) [inline examples/SluiceTest/sluice_tester.py:98]: you shouldn't need this if you look at build -- you can look at the class of the layer and figure it out from the DAG scan in TG.   Don't worry about this until we get some good results or are ready to merge to deepchem/deepchem though

File: deepchem/models/tensorgraph/layers.py [hunk 8]
```diff
@@ -140,6 +140,129 @@ def convert_to_layers(in_layers):
...
+
+class SluiceLoss(Layer):
+    """
+    Calculates the loss in a Sluice Network
+    """
+
+    def __init__(self, **kwargs):
+        super(SluiceLoss, self).__init__(**kwargs)
+
+    def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py:239]: are you sure you are calculating this right?  tests?
I think that this loss is not weighted correctly/is overpowering the weighted error.

The only purpose of this is for towers to not calculate the same features.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 9]
```diff
@@ -19,610 +21,647 @@
 
 class TensorGraph(Model):
 
-  def __init__(self,
```
Comment (wcjordan) [inline deepchem/models/tensorgraph/tensor_graph.py:22]: Any tips for reading this file's changeset?  What's changed here?

Comment (galenxing) [reply deepchem/models/tensorgraph/tensor_graph.py:22]: Didn't change too much. Just added some crude methods so that it printed out my beta and alpha share values at given epochs. The majority of the diff came from weird formatting.

Comment (wcjordan) [reply deepchem/models/tensorgraph/tensor_graph.py:22]: Cool, could you comment on those methods for printing the share values to point them out?
